src/lib/swapper/swappers/CowSwapper/cowBuildTrade/cowBuildTrade.ts

```diff
@@ -1,84 +1,50 @@
 import { Err, Ok } from '@sniptt/monads'
 import { bn, bnOrZero } from 'lib/bignumber/bignumber'
 import { fromBaseUnit, toBaseUnit } from 'lib/math'
 import type { BuildTradeInput, SwapErrorRight } from 'lib/swapper/api'
-import { makeSwapErrorRight, SwapErrorType } from 'lib/swapper/api'
-import type { CowSwapperDeps } from 'lib/swapper/swappers/CowSwapper/CowSwapper'
+import type { CowChainId, CowSwapperDeps } from 'lib/swapper/swappers/CowSwapper/CowSwapper'
 import type { CowSwapQuoteResponse, CowTrade } from 'lib/swapper/swappers/CowSwapper/types'
 import {
-  COW_SWAP_ETH_MARKER_ADDRESS,
   DEFAULT_APP_DATA,
   DEFAULT_SOURCE,
   ORDER_KIND_SELL,
 } from 'lib/swapper/swappers/CowSwapper/utils/constants'
 import { cowService } from 'lib/swapper/swappers/CowSwapper/utils/cowService'
-import { getNowPlusThirtyMinutesTimestamp } from 'lib/swapper/swappers/CowSwapper/utils/helpers/helpers'
+import {
+  getCowswapNetwork,
+  getNowPlusThirtyMinutesTimestamp,
+} from 'lib/swapper/swappers/CowSwapper/utils/helpers/helpers'
 import {
   selectBuyAssetUsdRate,
   selectSellAssetUsdRate,
 } from 'state/zustand/swapperStore/amountSelectors'
 import { swapperStore } from 'state/zustand/swapperStore/useSwapperStore'
-import {
-  convertDecimalPercentageToBasisPoints,
-  subtractBasisPointAmount,
-} from 'state/zustand/swapperStore/utils'
-
-export async function cowBuildTrade(
-  deps: CowSwapperDeps,
-  input: BuildTradeInput,
-): Promise<Result<CowTrade<KnownChainIds.EthereumMainnet>, SwapErrorRight>> {
-  const { sellAsset, buyAsset, accountNumber, receiveAddress, slippage } = input
 
-  if (!receiveAddress)
-    return Err(
-      makeSwapErrorRight({
-        message: 'Receive address is required to build CoW trades',
-        code: SwapErrorType.MISSING_INPUT,
-      }),
-    )
+import { validateBuildTrade } from '../utils/validator'
 
-  const sellAmountBeforeFeesCryptoBaseUnit = input.sellAmountBeforeFeesCryptoBaseUnit
-
-  const { assetReference: sellAssetErc20Address, assetNamespace: sellAssetNamespace } = fromAssetId(
-    sellAsset.assetId,
-  )
+export async function cowBuildTrade<T extends CowChainId>(
+  { adapter, baseUrl }: CowSwapperDeps,
+  input: BuildTradeInput,
+): Promise<Result<CowTrade<T>, SwapErrorRight>> {
+  const { accountNumber, sellAsset, buyAsset } = input
 
-  const { assetReference: buyAssetErc20Address, chainId: buyAssetChainId } = fromAssetId(
-    buyAsset.assetId,
-  )
+  const network = getCowswapNetwork(adapter)
+  if (network.isErr()) return Err(network.unwrapErr())
 
-  if (sellAssetNamespace !== 'erc20') {
-    return Err(
-      makeSwapErrorRight({
-        message: '[cowBuildTrade] - Sell asset needs to be ERC-20 to use CowSwap',
-        code: SwapErrorType.UNSUPPORTED_PAIR,
-        details: { sellAssetNamespace },
-      }),
-    )
-  }
+  const maybeValidTradePair = validateBuildTrade(input)
+  maybeValidTradePair.isErr() && Err(maybeValidTradePair.unwrapErr())
```

Comment:
This doesn't do anything - will just evaluate to an `Err()` expression on the right side, we want to return this - ditto for all the `&& Err()` occurrences.